### PR TITLE
Fix ZeRO-3 + PEFT + gradient checkpointing

### PR DESCRIPTION
**Pull request by qgallouedec**
This PR does 4 things:

### 1. Move  the `use_reentrant` default overriding inside the trainers

just before calling `super().__init__`. This is required by (2) because we need to know during the init if the user has explicitly pass `use_reentrant=False` or not to raise a warning

### 2. When PEFT is used with DeepSpeed ZeRO-3 and gradient checkpointing in SFTTrainer, force using reentrant

When `use_reentrant` is set to `False` **by the user**, it also raises a warnings:

```
You are using PEFT with DeepSpeed ZeRO-3 and gradient checkpointing with `use_reentrant=False`.
`use_reentrant` is forced to `True` in this configuration to ensure correct training. To remove
this warning, unset `use_reentrant` in `gradient_checkpointing_kwargs` or set it to `True`.
```

### 3. Update the xfail reason of `TestDistributed::test_sft_peft[zero2]`

Now, `TestDistributed::test_sft_peft[zero2]` still fails, but for the same reason as `TestDistributed::test_sft_peft[zero3]`

### 4. Update the xfail condition of `TestDistributed::test_sft_peft[zero2,zero3]` to torch 2.10 only

After investigation, `TestDistributed::test_sft_peft[zero2,zero3]` are failing because of torch 2.10 specifically. For more information see https://github.com/huggingface/transformers/pull/43692.

### CODE REVIEW

**Review by albertvillanova — APPROVED** [review]
Thanks. I thinks this change is because we also accept TrainingArguments as args.
